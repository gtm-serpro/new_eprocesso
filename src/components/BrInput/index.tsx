import Icon from "@/common/Icon";
import classNames from "classnames";
import React from "react";
import { NumericFormat, PatternFormat } from "react-number-format";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import useUniqueId from "../../hooks/useUniqueId";
import IMtProps from "../../types/IMtProps";
import Message from "../BrMessage";

export interface BrInputProps extends React.InputHTMLAttributes<HTMLInputElement>, IMtProps {
  /**
   * Label do input.
   */
  label?: React.ReactNode;
  /**
   * Tamanho do input.
   * - small: pequeno
   * - medium: médio (padrão)
   * - large: grande
   */
  density?: "small" | "medium" | "large";
  /**
   * Nome do ícone Font Awesome do input.
   */
  icon?: string;
  /**
   * Botão no canto direito do input.
   */
  button?: React.ReactNode;
  /**
   * Se é do tipo "highlight".
   */
  highlight?: boolean;
  /**
   * Se a label fica na lateral.
   */
  inline?: boolean;
  /**
   * Input útil para valores numéricos formatados (ex: moeda), bem como
   * para valores sem formatação específica, mas que só aceitam números (ex: números
   * de documentos que não possuem uma formatação definida).
   *
   * O valor da prop `numeric` pode ser um objeto ou um boolean. Quando é um objeto,
   * aplica propriedades específicas para formatação de valores numéricos. Quando é
   * boolean `true` ou um objeto vazio `{}`, aplica as propriedades padrão (ver
   * referência abaixo).
   *
   * Propriedades do objeto `numeric`:
   * <https://s-yadav.github.io/react-number-format/docs/numeric_format/>
   *
   * Alguns valores padrão foram modificados para atender a uma maior quantidade de
   * casos de uso da realidade brasileira:
   *
   * - `allowNegative`: `false`
   * - `allowLeadingZeros`: `true`
   * - `decimalScale`: `0`
   * - `fixedDecimalScale`: `true`
   * - `decimalSeparator`: `,`
   */
  numeric?:
    | boolean
    | {
        thousandSeparator?: boolean | string;
        decimalSeparator?: string;
        allowedDecimalSeparators?: Array<string>;
        thousandsGroupStyle?: "thousand" | "lakh" | "wan" | "none";
        decimalScale?: number;
        allowNegative?: boolean;
        allowLeadingZeros?: boolean;
        fixedDecimalScale?: boolean;
        suffix?: string;
        prefix?: string;
      };
  /**
   * Status do input muda a sua cor.
   * - success: verde
   * - danger: vermelho
   * - info: azul
   * - warning: amarelo
   */
  status?: "success" | "danger" | "info" | "warning";
  /**
   * Texto de feedback que aparece embaixo do item, com a cor de fundo de acordo com o status escolhido.
   */
  feedbackText?: string;
  /**
   * Máscara para o input.
   */
  mask?: string;
  /**
   * Placeholder para a máscara. Padrão: "_"
   */
  maskPlaceholder?: string;
  /**
   * Se deve mostrar o placeholder da máscara. Padrão: true
   */
  showMaskPlaceholder?: boolean;
}

const BrInput = React.forwardRef<HTMLInputElement, BrInputProps>(
  (
    {
      className,
      id,
      children,
      numeric,
      label,
      density = "medium",
      icon,
      button,
      highlight = false,
      inline = false,
      status,
      feedbackText,
      mask,
      maskPlaceholder = "_",
      showMaskPlaceholder = true,
      ...props
    },
    ref
  ) => {
    const fid = useUniqueId(id, "input_____");
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);

    const commonInputProps = {
      id: fid,
      className: classNames(icon && "has-icon"),
      ...spreadProps,
    };

    return (
      <div
        className={classNames(
          "br-input",
          density === "small" && "small",
          density === "large" && "large",
          highlight && "input-highlight",
          inline && "input-inline",
          status,
          className,
          ...mtProps
        )}
      >
        {label && (
          <div className="input-label">
            <label htmlFor={fid}>{label}</label>
          </div>
        )}
        <div className="input-content">
          <div className="input-group">
            {icon && (
              <div className="input-icon">
                <Icon icon={icon} />
              </div>
            )}
            {!!numeric ? (
              <NumericFormat
                getInputRef={ref}
                {...commonInputProps}
                allowNegative={false}
                allowLeadingZeros
                decimalScale={0}
                fixedDecimalScale
                decimalSeparator=","
                {...(typeof numeric === "object" && numeric)}
              />
            ) : !!mask ? (
              <PatternFormat
                getInputRef={ref}
                format={mask}
                {...(showMaskPlaceholder && { mask: maskPlaceholder })}
                {...commonInputProps}
                onChange={() => null}
                onValueChange={({ formattedValue }, { event }) => {
                  commonInputProps.onChange?.({
                    ...event,
                    target: {
                      ...event?.target,
                      value: formattedValue.replaceAll(maskPlaceholder, ""),
                    },
                  });
                }}
              />
            ) : (
              <input ref={ref} {...commonInputProps} />
            )}
            {button}
            {feedbackText &&
              (status ? (
                <Message
                  category="feedback"
                  status={status ? status : "success"}
                  message={feedbackText}
                />
              ) : (
                <p className="text-down-01">{feedbackText}</p>
              ))}
            {children}
          </div>
        </div>
      </div>
    );
  }
);

export default BrInput;
