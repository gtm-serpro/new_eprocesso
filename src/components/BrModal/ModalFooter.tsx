import BrButton from "@/components/BrButton";
import classNames from "classnames";
import { MouseEventHandler, useEffect, useRef } from "react";
import { BrModalActions } from ".";

type ButtonProps = {
  type?: "submit" | "reset" | "button";
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const ModalFooter = ({
  primaryAction,
  secondaryAction,
  keepOpenOnPrimaryAction,
  keepOpenOnSecondaryAction,
  focusOnPrimaryAction,
  focusOnSecondaryAction,
  onClose,
  formId,
}: BrModalActions) => {
  const justify = !!primaryAction && !!secondaryAction ? "end" : "center";
  const classes = classNames("br-modal-footer px-2", `justify-content-${justify}`);

  const secondaryButtonRef = useRef<HTMLButtonElement>(null);
  const primaryButtonRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    if (focusOnPrimaryAction) {
      if (primaryButtonRef.current && document.activeElement !== primaryButtonRef.current) {
        primaryButtonRef.current.contentEditable = "true";
        primaryButtonRef.current.focus?.();
        primaryButtonRef.current.contentEditable = "false";
      }
    } else if (focusOnSecondaryAction) {
      if (secondaryButtonRef.current && document.activeElement !== secondaryButtonRef.current) {
        secondaryButtonRef.current.contentEditable = "true";
        secondaryButtonRef.current?.focus?.();
        secondaryButtonRef.current.contentEditable = "false";
      }
    }
  }, [primaryButtonRef, secondaryButtonRef]);

  const handlePrimary = () => {
    primaryAction?.action && primaryAction.action();
    !keepOpenOnPrimaryAction && onClose?.();
  };

  const handleSecondary = () => {
    secondaryAction?.action && secondaryAction.action();
    !keepOpenOnSecondaryAction && onClose?.();
  };

  const buttonProps: ButtonProps = formId
    ? { type: "submit" }
    : { type: "button", onClick: handlePrimary };

  return primaryAction || secondaryAction ? (
    <div className={classes}>
      {!!secondaryAction && (
        <BrButton
          ref={secondaryButtonRef}
          secondary
          size="small"
          mt={1}
          ml={1}
          onClick={handleSecondary}
          disabled={secondaryAction.disabled}
          block
          clearBlock="md"
        >
          {secondaryAction.label || "Cancelar"}
        </BrButton>
      )}
      {!!primaryAction && (
        <BrButton
          ref={primaryButtonRef}
          primary
          size="small"
          mt={1}
          ml={1}
          disabled={primaryAction.disabled}
          block
          clearBlock="md"
          {...buttonProps}
        >
          {primaryAction.label || "Confirmar"}
        </BrButton>
      )}
    </div>
  ) : null;
};

export default ModalFooter;
