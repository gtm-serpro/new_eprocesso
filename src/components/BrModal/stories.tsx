import BrButton from "@/components/BrButton";
import { Meta, Story } from "@storybook/react";
import { useState } from "react";

import BrModal, { BrModalProps } from ".";

export default {
  title: "Superfícies/Modal",
  component: BrModal,
} as Meta;

const Template: Story<BrModalProps> = (args) => {
  const [open, setOpen] = useState(false);

  return (
    <>
      <BrButton primary onClick={() => setOpen(true)}>
        Abrir Modal
      </BrButton>
      <BrModal {...args} isOpen={open} onClose={() => setOpen(false)}>
        Elementos que compõem o conteúdo do modal. Definem a largura e a altura do mesmo.
      </BrModal>
    </>
  );
};

export const Simples = Template.bind({});

export const ComActionPrincipal = Template.bind({});
ComActionPrincipal.args = {
  title: "Título do Modal",
  primaryAction: { label: "Ok", action: () => "", disabled: true },
};

export const ComActions = Template.bind({});
ComActions.args = {
  title: "Título do Modal",
  primaryAction: { action: () => "" },
  secondaryAction: { action: () => "" },
};
