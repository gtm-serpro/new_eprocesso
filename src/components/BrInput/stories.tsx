import { Meta, Story } from "@storybook/react";

import BrInput, { BrInputProps } from ".";

import { Col, Row } from "@/common/Grid";

export default {
  title: "Formulários/Input",
  component: BrInput,
  // argTypes: {
  //   numericProps: {
  //     control: "object",
  //   },
  //   status: {
  //     control: "radio",
  //     options: {
  //       default: "",
  //       success: "success",
  //       danger: "danger",
  //       info: "info",
  //       warning: "warning",
  //     },
  //   },
  // },
} as Meta;

const Template: Story<BrInputProps> = (args) => (
  <Row>
    <Col sm={3}>
      <BrInput {...args} />
    </Col>
  </Row>
);

export const Simples = Template.bind({});
Simples.args = {
  label: "Label",
  placeholder: "Placeholder customizado",
};

export const Desabilitado = Template.bind({});
Desabilitado.args = {
  label: "Input desabilitado",
  placeholder: "Placeholder customizado",
  disabled: true,
};

export const Numerico = Template.bind({});
Numerico.storyName = "Numérico";
Numerico.args = {
  placeholder: "Digite apenas os números do seu RG",
  label: "RG",
  numeric: true,
  maxLength: 20,
};

export const Formatado = Template.bind({});
Formatado.storyName = "Numérico formatado";
Formatado.args = {
  label: "Input de número",
  placeholder: "Digite um valor em reais",
  feedbackText:
    "É preciso digitar a vírgula ou mover o cursor para após a vírgula para separar os centavos.",
  maxLength: 20,
  numeric: {
    decimalScale: 2,
    prefix: "R$ ",
  },
};

export const Mask = Template.bind({});
Mask.storyName = "Com máscara";
Mask.args = {
  label: "Digite o seu CPF",
  mask: "###.###.###-##",
};

export const ComFeedback = Template.bind({});
ComFeedback.args = {
  label: "Label Customizado",
  placeholder: "Placeholder customizado",
  status: "danger",
  feedbackText: "Campo obrigatório",
};
