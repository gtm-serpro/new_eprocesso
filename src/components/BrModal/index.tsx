import Icon from "@/common/Icon";
import BrButton from "@/components/BrButton";
import { useBodyOverflowHidden } from "@/hooks/useBodyOverflowHidden";
import { useWindowWidth } from "@/hooks/useWindowWidth";
import classNames from "classnames";
import { Fragment, MouseEvent } from "react";
import ModalFooter from "./ModalFooter";

type ModalAction = {
  label?: string;
  disabled?: boolean;
  action?: () => void;
};

export interface BrModalState {
  isOpen: boolean;
  showClose?: boolean;
}

export interface BrModalActions {
  formId?: string;
  keepOpenOnPrimaryAction?: boolean;
  keepOpenOnSecondaryAction?: boolean;
  focusOnPrimaryAction?: boolean;
  focusOnSecondaryAction?: boolean;
  primaryAction?: ModalAction;
  secondaryAction?: ModalAction;
  onClose?: () => void;
}

type dimension = number | string;

export interface BrModalProps extends BrModalState, BrModalActions {
  children?: React.ReactNode;
  title?: string;
  minWidth?: dimension;
  minHeight?: dimension;
  height?: dimension;
  width?: dimension;
  formId?: string;
  primaryAction?: ModalAction;
  secondaryAction?: ModalAction;
}

const BrModal = (props: BrModalProps) => {
  const {
    children,
    title,
    showClose,
    primaryAction,
    secondaryAction,
    isOpen,
    keepOpenOnPrimaryAction,
    keepOpenOnSecondaryAction,
    focusOnPrimaryAction,
    focusOnSecondaryAction,
    onClose,
    minWidth,
    minHeight,
    height,
    width,
    formId,
  } = props;

  const scrimClasses = classNames("br-scrim", "foco", { active: true });

  const isSmallScreen = useWindowWidth() < 1024;

  const handleClickOnScrim = (e: MouseEvent<HTMLDivElement>) => {
    if (e.target === e.currentTarget) {
      onClose?.();
    }
  };

  const Wrapper = formId ? "form" : Fragment;
  const wrapperProps = formId ? { id: formId, onSubmit: primaryAction?.action } : {};

  useBodyOverflowHidden(isOpen);

  if (!isOpen) {
    return null;
  }

  return (
    <Wrapper {...wrapperProps}>
      <div className={scrimClasses} onClick={handleClickOnScrim}>
        <div className="br-modal p-0 auto">
          <div className="br-card py-2" style={{ flexDirection: "column" }}>
            <div className="br-modal-header">
              {title && (
                <div className="br-modal-title px-2" title={title}>
                  {title}
                </div>
              )}
              {(showClose || title) && (
                <BrButton circle close size="small" aria-label="Fechar" onClick={onClose}>
                  <Icon icon="times" aria-hidden="true" />
                </BrButton>
              )}
            </div>
            <div
              className="mb-2 px-2"
              style={{
                overflowY: "auto",
                width: isSmallScreen ? "95vw" : width,
                minWidth: isSmallScreen ? "95vw" : minWidth,
                height,
                minHeight,
                maxHeight: "calc(100vh - 200px)",
              }}
            >
              {children}
            </div>
            <ModalFooter
              formId={formId}
              primaryAction={primaryAction}
              secondaryAction={secondaryAction}
              keepOpenOnPrimaryAction={keepOpenOnPrimaryAction}
              keepOpenOnSecondaryAction={keepOpenOnSecondaryAction}
              focusOnPrimaryAction={focusOnPrimaryAction}
              focusOnSecondaryAction={focusOnSecondaryAction}
              onClose={onClose}
            />
          </div>
        </div>
      </div>
    </Wrapper>
  );
};

export default BrModal;
