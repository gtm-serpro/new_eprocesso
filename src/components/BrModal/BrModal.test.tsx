import { fireEvent, render, screen } from "@testing-library/react";
import BrInput from "../BrInput";

import BrModal from ".";

describe("<BrModal />", () => {
  it("deveria renderizar modal com conteúdo passado", () => {
    const { container } = render(
      <BrModal isOpen={true} onClose={() => null}>
        <BrInput label="Conteúdo do Modal" />
      </BrModal>
    );

    screen.getByLabelText(/conteúdo do modal/i);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("deveria renderizar modal com título se informado", () => {
    render(
      <BrModal isOpen={true} onClose={() => null} title="Título">
        Conteúdo
      </BrModal>
    );

    const titleElement = screen.getByTitle(/título/i);
    expect(titleElement).toHaveClass("br-modal-title");
    expect(titleElement.parentElement).toHaveClass("br-modal-header");
  });

  it("deveria renderizar modal com botão fechar se existir título", () => {
    render(
      <BrModal isOpen={true} onClose={() => null} title="Título">
        Conteúdo
      </BrModal>
    );

    const closeButton = screen.getByLabelText(/fechar/i);
    expect(closeButton).toHaveClass("br-button small");
  });

  it('deveria chamar "onClose" modal se botão "fechar" acionado', () => {
    const handleClose = jest.fn();
    const { getByLabelText } = render(
      <BrModal isOpen={true} onClose={handleClose} title="Título">
        Conteúdo
      </BrModal>
    );

    fireEvent.click(getByLabelText(/fechar/i));
    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  it('deveria chamar "onClose" modal scrim for clicado', () => {
    const handleClose = jest.fn();
    const { container } = render(
      <BrModal isOpen={true} onClose={handleClose} title="Título">
        Conteúdo
      </BrModal>
    );

    const element = container.firstElementChild || new Element();
    fireEvent.click(element);
    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  it("deveria exibir ação primária centralizada", () => {
    render(
      <BrModal
        isOpen={true}
        onClose={() => null}
        primaryAction={{ label: "Confirmar", action: () => null }}
      >
        Conteúdo
      </BrModal>
    );

    const botaoPrimario = screen.getByText(/confirmar/i);
    expect(botaoPrimario.parentElement).toHaveClass("br-modal-footer justify-content-center");
  });

  it("deveria exibir ação secundária centralizada", () => {
    render(
      <BrModal
        isOpen={true}
        onClose={() => null}
        secondaryAction={{ label: "Cancelar", action: () => null }}
      >
        Conteúdo
      </BrModal>
    );

    const botaoSecundario = screen.getByText(/cancelar/i);
    expect(botaoSecundario.parentElement).toHaveClass("br-modal-footer justify-content-center");
  });

  it("deveria exibir ambas as ações alinhadas à direita", () => {
    render(
      <BrModal
        isOpen={true}
        onClose={() => null}
        primaryAction={{ label: "Ok", action: () => null }}
        secondaryAction={{ label: "Cancel", action: () => null }}
      >
        Conteúdo
      </BrModal>
    );

    const botaoPrimario = screen.getByText(/ok/i);
    expect(botaoPrimario.parentElement).toHaveClass("br-modal-footer justify-content-end");
    const botaoSecundario = screen.getByText(/cancel/i);
    expect(botaoSecundario.parentElement).toHaveClass("br-modal-footer justify-content-end");
  });

  it("deveria exibir ambas as ações com texto padrão, se não informado", () => {
    render(
      <BrModal
        isOpen={true}
        onClose={() => null}
        primaryAction={{ action: () => null }}
        secondaryAction={{ action: () => null }}
      >
        Conteúdo
      </BrModal>
    );

    screen.getByText(/confirmar/i);
    screen.getByText(/cancelar/i);
  });

  it("deveria fechar modal se ação primária executada", () => {
    const handleClose = jest.fn();
    const handleConfirmar = jest.fn();

    render(
      <BrModal
        isOpen={true}
        onClose={handleClose}
        primaryAction={{ label: "Confirmar", action: handleConfirmar }}
      >
        Conteúdo
      </BrModal>
    );

    fireEvent.click(screen.getByText(/confirmar/i));
    expect(handleClose).toHaveBeenCalledTimes(1);
    expect(handleConfirmar).toHaveBeenCalledTimes(1);
  });

  it("deveria fechar modal se ação secundária executada", () => {
    const handleClose = jest.fn();
    const handleCancelar = jest.fn();

    render(
      <BrModal
        isOpen={true}
        onClose={handleClose}
        secondaryAction={{ label: "Cancelar", action: handleCancelar }}
      >
        Conteúdo
      </BrModal>
    );

    fireEvent.click(screen.getByText(/cancelar/i));
    expect(handleClose).toHaveBeenCalledTimes(1);
    expect(handleCancelar).toHaveBeenCalledTimes(1);
  });
});
