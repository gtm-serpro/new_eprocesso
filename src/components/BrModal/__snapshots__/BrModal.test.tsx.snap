// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrModal /> deveria renderizar modal com conteúdo passado 1`] = `
<div
  class="br-scrim foco active"
>
  <div
    class="br-modal p-0 auto"
  >
    <div
      class="br-card py-2"
      style="flex-direction: column;"
    >
      <div
        class="br-modal-header"
      />
      <div
        class="mb-2 px-2"
        style="overflow-y: auto; max-height: calc(100vh - 200px);"
      >
        <div
          class="br-input"
        >
          <div
            class="input-label"
          >
            <label
              for="input_____1"
            >
              Conteúdo do Modal
            </label>
          </div>
          <div
            class="input-content"
          >
            <div
              class="input-group"
            >
              <input
                class=""
                id="input_____1"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
